File: lib/Karve/CustomControls/DataGrid/Clases/DataGridRules/DataGridRules.vb
Public Class DataGridRules
    Inherits System.Collections.Hashtable

    Public Overloads Sub Add(ByVal TablaQuery As DataGridRule)
        TablaQuery.Item = Me.Count
        MyBase.Add(TablaQuery.Name, TablaQuery)
    End Sub

    Public Overloads Sub Remove(ByVal Key As Object)
        MyBase.Remove(Key)
        Dim i As Integer = 0
        Dim HC = (From c As DataGridRule In Me.Values Order By c.Item Select c)
        Dim RS As New ArrayList
        For Each ctr As DataGridRule In HC
            ctr.Item = i
            i += 1
        Next
    End Sub

    Public Function Order() As ArrayList
        Dim HC = (From c As DataGridRule In Me.Values Order By c.Item Select c)
        Dim RS As New ArrayList
        For Each ctr As DataGridRule In HC
            RS.Add(ctr)
        Next
        Return RS
    End Function

    Public Function Existe(Value As String) As Boolean
        Return Me.ContainsKey(Value)
    End Function

    Public Function DameItemExiste(Value As String) As DataGridRule
        Dim HC = (From c As DataGridRule In Me.Values Where c.Name = Value Order By c.Item Select c)
        If HC.Count <> 0 Then
            Return HC(0)
        Else : Return Nothing
        End If
    End Function

    Public Sub Modify(ByVal TablaQuery As DataGridRule)
        MyBase.Item(TablaQuery.Name) = TablaQuery
    End Sub

    Public Function DameWhere() As String
        Dim _sWhere As String = ""
        Dim RL As New DataGridRule
        Dim _sClausula As String = ""
        Dim _sCampo As String = ""
        Dim _sOperador As String = ""
        '*====================================================================
        For Each RL In Order()
            If RL.Item > 0 Then _sOperador = RL.OperadorVal
            If RL.Expresion <> "" Then
                _sClausula = RL.Expresion
            Else
                _sClausula = DameClausula(RL)
            End If
            _sWhere = _sWhere & _sOperador & _sClausula
        Next
        If _sWhere <> "" Then _sWhere = " WHERE " & _sWhere
        Return _sWhere
    End Function

    Public Function DameClausula(RL As DataGridRule) As String
        Dim _sClausula As String = ""
        Dim _sCampo As String = ""

        _sCampo = ""

        If RL.ExpresionBD <> "" Then
            _sCampo = RL.ExpresionBD
        ElseIf RL.AliasCampo <> "" And RL.AliasCampo <> RL.Campo Then
            _sCampo = RL.AliasCampo
        Else
            _sCampo = RL.Tabla & "."
            _sCampo += RL.Campo
        End If

        If RL.Criterio = DataGridRule.euCriterio.Contiene Then
            _sClausula = _sCampo & RL.CriterioVal & "'%" & RL.Valor & "%'"
        ElseIf RL.Criterio = DataGridRule.euCriterio.NoContiene Then
            _sClausula = _sCampo & RL.CriterioVal & "'%" & RL.Valor & "%'"
        ElseIf RL.Criterio = DataGridRule.euCriterio.Distinto Then
            _sClausula = _sCampo & RL.CriterioVal & "'" & RL.Valor & "'"
        ElseIf RL.Criterio = DataGridRule.euCriterio.Igual Then
            _sClausula = _sCampo & RL.CriterioVal & "'" & RL.Valor & "'"
        ElseIf RL.Criterio = DataGridRule.euCriterio.Empieza Then
            _sClausula = _sCampo & RL.CriterioVal & "'" & RL.Valor & "%'"
        ElseIf RL.Criterio = DataGridRule.euCriterio.Termina Then
            _sClausula = _sCampo & RL.CriterioVal & "'%" & RL.Valor & "'"
        ElseIf RL.Criterio = DataGridRule.euCriterio.EsNulo Then
            _sClausula = _sCampo & " IS NULL "
        ElseIf RL.Criterio = DataGridRule.euCriterio.NoEsNulo Then
            _sClausula = _sCampo & " IS NOT NULL "
        Else
            _sClausula = _sCampo & RL.CriterioVal & "'" & RL.Valor & "'"
        End If

        Return _sClausula
    End Function

End Class

File: src/Controls/KarveControls/DataGrid/DataGridHelpers/DataGridRules/DataGridRules.cs

using System.Collections;
using System.Collections.Generic;
using System.Linq;

public class DataGridRules : Dictionary<object, DataGridRule>
{

	public void Add(DataGridRule TablaQuery)
	{
		TablaQuery.Item = this.Count;
		base.Add(TablaQuery.Name, TablaQuery);
	}

    public void Remove(object Key)
    {
        base.Remove(Key);
        int i = 0;
        dynamic HC = from c in Values orderby c.Item select c;
        ArrayList RS = new ArrayList();
        foreach (DataGridRule ctr in HC)
        {
            ctr.Item = i;
            i += 1;
        }
    }

    public IList<DataGridRule> Order()
	{
		dynamic HC = from c in Values orderby c.Item select c;
		IList<DataGridRule> RS = new List<DataGridRule>();
		foreach (DataGridRule ctr in HC) {
			RS.Add(ctr);
		}
		return RS;
	}

	public bool Existe(string Value)
	{
		return this.ContainsKey(Value);
	}

	public DataGridRule DameItemExiste(string Value)
	{
		dynamic HC = (from c in Values where c.Name == Value orderby c.Item select c);
		if (HC.Count != 0) {
			return HC(0);
		} else {
			return null;
		}
	}

	public void Modify(DataGridRule TablaQuery)
	{
        
        base[TablaQuery.Name] = TablaQuery;
	}

	public string DameWhere()
	{
		string _sWhere = "";
		DataGridRule RL = new DataGridRule();
		string _sClausula = "";
		string _sCampo = "";
		string _sOperador = "";
	    IList<DataGridRule> orders = Order();
		//*====================================================================
		foreach (DataGridRule rule in orders) {
			if (rule.Item > 0)
				_sOperador =rule.OperadorVal;
			if (!string.IsNullOrEmpty(rule.Expresion)) {
				_sClausula = rule.Expresion;
			} else {
				_sClausula = DameClausula(rule);
			}
			_sWhere = _sWhere + _sOperador + _sClausula;
		}
		if (!string.IsNullOrEmpty(_sWhere))
			_sWhere = " WHERE " + _sWhere;
		return _sWhere;
	}

	public string DameClausula(DataGridRule RL)
	{
		string _sClausula = "";
		string _sCampo = "";

		_sCampo = "";

		if (!string.IsNullOrEmpty(RL.ExpresionBD)) {
			_sCampo = RL.ExpresionBD;
		} else if (!string.IsNullOrEmpty(RL.AliasCampo) & RL.AliasCampo != RL.Campo) {
			_sCampo = RL.AliasCampo;
		} else {
			_sCampo = RL.Tabla + ".";
			_sCampo += RL.Campo;
		}

		if (RL.Criterio == DataGridRule.euCriterio.Contiene) {
			_sClausula = _sCampo + RL.CriterioVal + "'%" + RL.Valor + "%'";
		} else if (RL.Criterio == DataGridRule.euCriterio.NoContiene) {
			_sClausula = _sCampo + RL.CriterioVal + "'%" + RL.Valor + "%'";
		} else if (RL.Criterio == DataGridRule.euCriterio.Distinto) {
			_sClausula = _sCampo + RL.CriterioVal + "'" + RL.Valor + "'";
		} else if (RL.Criterio == DataGridRule.euCriterio.Igual) {
			_sClausula = _sCampo + RL.CriterioVal + "'" + RL.Valor + "'";
		} else if (RL.Criterio == DataGridRule.euCriterio.Empieza) {
			_sClausula = _sCampo + RL.CriterioVal + "'" + RL.Valor + "%'";
		} else if (RL.Criterio == DataGridRule.euCriterio.Termina) {
			_sClausula = _sCampo + RL.CriterioVal + "'%" + RL.Valor + "'";
		} else if (RL.Criterio == DataGridRule.euCriterio.EsNulo) {
			_sClausula = _sCampo + " IS NULL ";
		} else if (RL.Criterio == DataGridRule.euCriterio.NoEsNulo) {
			_sClausula = _sCampo + " IS NOT NULL ";
		} else {
			_sClausula = _sCampo + RL.CriterioVal + "'" + RL.Valor + "'";
		}

		return _sClausula;
	}

}